src/tools/schema2base.ts

```diff
@@ -25,14 +25,58 @@ export type AddFieldOptions = Readonly<{
   isCollection?: boolean;
 }>;
 
-export interface ClassGenerator {
-  addField(opts: AddFieldOptions): void;
-  generate(fieldCount: number): string;
+export interface EntityGenerator {
+  generate(): string;
 }
 
 export class NodeAndGenerator {
   node: SchemaNode;
-  generator: ClassGenerator;
+  generator: EntityGenerator;
+}
+
+/**
+ * Iterates over schema fields and composes metadata useful for entity codegen.
+ */
+export abstract class SchemaDescriptorBase {
+
+  constructor(readonly node: SchemaNode) {}
+
+  process() {
+    for (const [field, descriptor] of Object.entries(this.node.schema.fields)) {
```

Comment:
This is an excellent place to have this block.

Reply:
Well... no, I don't think we have explicit unit tests for this. But it's not obvious to me we should have them. You want to check the behaviour of the component, so I would rather have us have tests for what Kotlin code we generate. If you put too much tests around the implementation details that you will make it a huge pain in the ass to do refactorings like this PR. While having tests for the output of the code generator allows making internal changes and having good confidence that the external behaviour is unchanged - which is what we want.

Reply:
Maybe one more thing that I have not said, but I should have: I don't consider the SchemaDescriptor to be a self contained piece of functionality. I see it as an internal structure of the code generator. Which is why I would argue against testing it in isolation. Especially that in it's not decoupled from the rest.

Schema2Graph is an interesting example where my judgement is on the other side - it is an internal detail of the code generator, but its function is well defined enough and not  that closely tied to the code generator that writing standalone unit tests (which can be seen as: writing a contract for its interface) is warranted.

With SchemaDescriptor I do not want to solidify its contract in the form of unit tests.